src/PolyType/GenerateShapeForAttribute.cs

```diff
@@ -0,0 +1,47 @@
+﻿using System.Diagnostics;
+
+namespace PolyType;
+
+/// <summary>
+/// Instructs the PolyType source generator to include <typeparamref name="T"/>
+/// in the <see cref="ITypeShapeProvider"/> that it generates.
+/// </summary>
+/// <typeparam name="T">The type for which shape metadata will be generated.</typeparam>
+/// <remarks>
+/// <para>
+/// The source generator will include a static property in the annotated class pointing
+/// to the <see cref="ITypeShapeProvider"/> that was generated for the entire project.
+/// </para>
+/// <para>
+/// For projects targeting .NET 8 or later, this additionally augments the class
+/// with an implementation of IShapeable for <typeparamref name="T"/>.
+/// </para>
+/// </remarks>
+/// <seealso cref="GenerateShapeForAttribute"/>
+/// <seealso cref="GenerateShapeAttribute"/>
+[AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct, AllowMultiple = true, Inherited = false)]
+[Conditional("NEVER")] // only the source generator uses this, and avoiding generic attributes avoid .NET Framework and Unity issues.
+public sealed class GenerateShapeForAttribute<T> : Attribute;
```

Comment:
Do we want to make this a subclass of the non-generic variant? Should make the reflection-based implementation simpler if anything.

Reply:
Reflection-based providers don't use this attribute _at all_. And source generators have to look for each attribute type specifically anyway. So no, it wouldn't really make anything simpler.